tests/components/apache_kafka/test_apache_kafka.py

```diff
@@ -0,0 +1,195 @@
+"""The tests for the Apache Kafka component."""
+from collections import namedtuple
+
+import pytest
+
+import homeassistant.components.apache_kafka as apache_kafka
+from homeassistant.core import split_entity_id
+from homeassistant.helpers.entityfilter import FILTER_SCHEMA
+from homeassistant.setup import async_setup_component
+
+from tests.async_mock import MagicMock, Mock, patch
+
+APACHE_KAFKA_PATH = "homeassistant.components.apache_kafka"
+
+
+@pytest.fixture(autouse=True, name="mock_client_methods")
+def mock_client_methods_fixture():
+    """Mock the apache kafka client."""
+    producer_path = f"{APACHE_KAFKA_PATH}.AIOKafkaProducer"
+    with patch(f"{producer_path}.start") as start:
+        with patch(f"{producer_path}.stop") as stop:
+            with patch(f"{producer_path}.send_and_wait") as send_and_wait:
+                yield (start, stop, send_and_wait)
+
+
+@pytest.fixture(autouse=True)
+def mock_bus_and_json(hass, monkeypatch):
+    """Mock the event bus listener and os component."""
+    hass.bus.async_listen = MagicMock()
+    monkeypatch.setattr(
+        f"{APACHE_KAFKA_PATH}.json.dumps", Mock(return_value=MagicMock())
+    )
+
+
+async def test_minimal_config(hass, mock_client_methods):
+    """Test the minimal config and defaults of component."""
+    config = {
+        apache_kafka.DOMAIN: {
+            "ip_address": "localhost",
+            "port": 8080,
+            "topic": "topic",
+        }
+    }
+    assert await async_setup_component(hass, apache_kafka.DOMAIN, config)
+
+
+async def test_full_config(hass, mock_client_methods):
+    """Test the full config of component."""
+    config = {
+        apache_kafka.DOMAIN: {
+            "ip_address": "localhost",
+            "port": 8080,
+            "topic": "topic",
+            "filter": {
+                "include_domains": ["light"],
+                "include_entity_globs": ["sensor.included_*"],
+                "include_entities": ["binary_sensor.included"],
+                "exclude_domains": ["light"],
+                "exclude_entity_globs": ["sensor.excluded_*"],
+                "exclude_entities": ["binary_sensor.excluded"],
+            },
+        }
+    }
+    assert await async_setup_component(hass, apache_kafka.DOMAIN, config)
+
+
+FilterTest = namedtuple("FilterTest", "id should_pass")
+
+
+def make_event(entity_id):
+    """Make a mock event for test."""
+    domain = split_entity_id(entity_id)[0]
+    state = MagicMock(
+        state="not blank",
+        domain=domain,
+        entity_id=entity_id,
+        object_id="entity",
+        attributes={},
+    )
+    return MagicMock(data={"new_state": state}, time_fired=12345)
+
+
+async def _setup(hass, filter_config):
+    """Shared set up for filtering tests."""
+    return apache_kafka.KafkaManager(
```

Comment:
We should use `async_setup_component` here, then `await hass.async_block_till_done()`

Mock out `send_and_wait` to capture the event.

And then fire mocked events.  Then we don't have to call any protected methods.

Reply:
So the issue I had with this is I don't know how to block until the event has finished processing so I know when to check to see if `send_and_wait` has been called. I tried this first but it didn't work:
```
event = make_event(test.id)
hass.bus.fire(apache_kafka.DOMAIN, event)
await hass.async_block_till_done()

assert test.should_pass == mock_client.send_and_wait.called_once
```
So then I tried adding `await hass.data[apache_kafka.DOMAIN].async_block_till_done()` like I saw in `logbook` but that didn't work either, just gave me an attribute error.

I think there's just a trick I'm missing. When I worked on `influxdb` the class had a nice `block_till_done` method I was able to use. Do I just need to add that to `KafkaManager`? 

Note that I was very hesitant to modify this file since there were no tests and I don't have a way to manually test it. But if its just adding a method I suppose that can't really do any harm.

Reply:
Also I wasn't sure if adding `block_till_done` was right anyway. Influx created its own thread and loop whereas this one just passes `hass.loop` when initializing the kafka client

Reply:
Does it work if you use `self.hass.bus.async_fire`  ?

Reply:
Actually its listening for `EVENT_STATE_CHANGED` so you can change the state of an entity once its setup as it should be listening.

` hass.states.async_set("fake.entity", STATE_ON)`

Reply:
Ok I think that did it but now I'm facing another issue, having trouble mocking the client correctly. Have to step out though so will continue in a bit.

Reply:
Ok got this one working. Mocking was a bit trickier then usual, patching the client itself didn't work. `AsyncMock` wasn't returning me awaitable methods. I think maybe its because kafka client is still using `asyncio.coroutine` but not really sure. Patching each method individually combined with using `async_set` like you suggested  got it working though 👍